indexing-service/src/main/java/io/druid/indexing/common/task/RealtimeIndexTask.java

```diff
@@ -349,15 +385,59 @@ public String getVersion(final Interval interval)
       }
     }
 
+    log.info("Job done!");
     return TaskStatus.success(getId());
   }
 
+  @Override
+  public boolean canRestore()
+  {
+    return true;
+  }
+
+  @Override
+  public void stopGracefully()
+  {
+    try {
+      synchronized (this) {
+        if (!stopped) {
+          stopped = true;
+          if (isFirehoseDrainableByClosing()) {
+            firehose.close();
+          }
+        }
+      }
+    }
+    catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Public for tests.
+   */
+  @JsonIgnore
+  public Firehose getFirehose()
+  {
+    return firehose;
+  }
+
   @JsonProperty("spec")
   public FireDepartment getRealtimeIngestionSchema()
   {
     return spec;
   }
 
+  /**
+   * Is our firehose drainable by closing it? If so, we should drain on stopGracefully rather than abruptly stopping.
+   * <p/>
+   * This is a hack to get around the fact that the Firehose interface does not help us do this.
+   */
+  private boolean isFirehoseDrainableByClosing()
+  {
+    return firehose instanceof EventReceiverFirehoseFactory.EventReceiverFirehose;
```

Comment:
yes, good point. Although, there are many other firehoses that can't be drained by closing, so I don't think that would work either. This method is tough to implement without help from the interface…